src/slicer.cpp

```diff
@@ -787,72 +787,59 @@ void SlicerLayer::makePolygons(const Mesh* mesh)
-    layers[0].z = std::max(0LL, initial_layer_thickness - thickness);
-    coord_t adjusted_layer_offset = initial_layer_thickness;
-    if (use_variable_layer_heights)
-    {
-        layers[0].z = adaptive_layers->at(0).z_position;
-    }
-    else if (slicing_tolerance == SlicingTolerance::MIDDLE)
-    {
-        layers[0].z = initial_layer_thickness / 2;
-        adjusted_layer_offset = initial_layer_thickness + (thickness / 2);
-    }
 
-    // define all layer z positions (depending on slicing mode, see above)
-    for (unsigned int layer_nr = 1; layer_nr < slice_layer_count; layer_nr++)
-    {
-        if (use_variable_layer_heights)
-        {
-            layers[layer_nr].z = adaptive_layers->at(layer_nr).z_position;
-        }
-        else
-        {
-            layers[layer_nr].z = adjusted_layer_offset + (thickness * (layer_nr - 1));
-        }
-    }
+    std::vector<std::pair<int32_t, int32_t>> zbbox = buildZHeightsForFaces(*mesh);
 
-    // loop over all mesh faces
-    for (unsigned int mesh_idx = 0; mesh_idx < mesh->faces.size(); mesh_idx++)
-    {
-        // get all vertices per face
-        const MeshFace& face = mesh->faces[mesh_idx];
-        const MeshVertex& v0 = mesh->vertices[face.vertex_index[0]];
-        const MeshVertex& v1 = mesh->vertices[face.vertex_index[1]];
-        const MeshVertex& v2 = mesh->vertices[face.vertex_index[2]];
 
-        // get all vertices represented as 3D point
-        Point3 p0 = v0.p;
-        Point3 p1 = v1.p;
-        Point3 p2 = v2.p;
+    buildSegments(*mesh, zbbox, layers);
 
-        // find the minimum and maximum z point
-        int32_t minZ = p0.z;
-        int32_t maxZ = p0.z;
-        if (p1.z < minZ) minZ = p1.z;
-        if (p2.z < minZ) minZ = p2.z;
-        if (p1.z > maxZ) maxZ = p1.z;
-        if (p2.z > maxZ) maxZ = p2.z;
+    log("slice of mesh took %.3f seconds\n", slice_timer.restart());
+
+    makePolygons(*mesh, slicing_tolerance, layers);
+    log("slice make polygons took %.3f seconds\n", slice_timer.restart());
+}
 
-        // calculate all intersections between a layer plane and a triangle
-        for (unsigned int layer_nr = 0; layer_nr < layers.size(); layer_nr++)
+void Slicer::buildSegments(const Mesh &mesh, const std::vector<std::pair<int32_t, int32_t>> &zbbox,
+    std::vector<SlicerLayer>& layers) 
+{
+    // OpenMP
+#pragma omp parallel for default(none) shared(mesh, zbbox, layers)
+    for (int layer_nr = 0; layer_nr < layers.size(); layer_nr++)
+    {
+        int32_t z = layers[layer_nr].z;
+        layers[layer_nr].segments.reserve(100);
+
+        // loop over all mesh faces
+        for (unsigned int mesh_idx = 0; mesh_idx < mesh.faces.size(); mesh_idx++)
         {
-            int32_t z = layers.at(layer_nr).z;
+            if ((z < zbbox[mesh_idx].first) || (z > zbbox[mesh_idx].second))
+                 continue;
```

Comment:
So you've now reversed this for loop from `faces -> layers` to `layers -> faces`. This requires you to build this `zbbox` mapping though in order to limit the layers that it's iterating over for each face. I don't see the point of this change; when the original loop nesting was used, it could keep the minimum and maximum Z coordinates local in the loop so it doesn't need to track them in a separate mapping.

Reply:
Enabling parallelization is the main point about reversing the loop.

Reply:
Why couldn't the parallelisation work over the original faces loop, then?

Reply:
Because you need write access on the layers. Thus it does not make much sense to parallize the faces loop. Because then the layer part must be "critical". 
In particular, the layer loop shall be "heavy" to boost up parallisation. Thus I have reversed the loops. Otherwise you parallize just over a few layers.

Reply:
Ah yes, I see. That's a good reason.